core/app/models/spree/gallery/product_gallery.rb

```diff
@@ -0,0 +1,16 @@
+module Spree
+  module Gallery
+    class ProductGallery
+      def initialize(product)
+        @product = product
+      end
+
+      # A list of all images associated with this gallery
+      #
+      # @return [Enumerable<Spree::Image>] all images in the gallery
+      def images
+        @images ||= (@product.images + @product.variant_images).uniq
```

Comment:
There is no need to merge these two collections, because `@product.variant_images` already [includes images of the master variant](https://github.com/solidusio/solidus/blob/master/core/app/models/spree/product.rb#L75) - what are [the images returned by `@product.images`](https://github.com/solidusio/solidus/blob/master/core/app/models/spree/product.rb#L72)